From 790dd56b2ff0965ed8d7c9a6f2c4cd45efeebd6c Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Mon, 23 Jun 2025 12:05:14 +0200
Subject: [PATCH 963/985] evl/net: dev: sanitize oob port
 activation/deactivation

The former interface EVL_SOCKIOC_{ACTIVATE/DEACTIVATE} simply could
not work, since one cannot bind an oob-enabled device which could be
returned by proto->getif() for activating it, without making that
device oob-enabled first.. Classic chicken-and-egg problem.

Rework the logic entirely, channeling request to open an oob port to
the 'net' root device, then use a fildes obtained on such port to
apply device-targeted requests, such as disabling it.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 include/evl/net/device.h          |  9 +--
 include/uapi/evl/net/device-abi.h |  1 +
 include/uapi/evl/net/net-abi.h    | 12 +++-
 include/uapi/evl/net/socket-abi.h |  7 ---
 kernel/evl/net/dev.c              | 99 +++++++++++++++++--------------
 kernel/evl/net/net.c              | 42 ++++++++++---
 kernel/evl/net/socket.c           | 13 +---
 7 files changed, 106 insertions(+), 77 deletions(-)

diff --git a/include/evl/net/device.h b/include/evl/net/device.h
index d7a0519f5..ef7c89f2e 100644
--- a/include/evl/net/device.h
+++ b/include/evl/net/device.h
@@ -13,13 +13,12 @@
 #include <uapi/evl/net/device-abi.h>
 #include <uapi/evl/net/bpf-abi.h>
 
-struct evl_socket;
-struct evl_netdev_activation;
+struct evl_net_devparams;
 struct notifier_block;
 struct sk_buff;
 
-int evl_net_switch_oob_port(struct evl_socket *esk,
-			    struct evl_netdev_activation *act);
+int evl_net_switch_oob_port(struct net_device *dev,
+			    struct evl_net_devparams *p);
 
 int evl_netdev_event(struct notifier_block *ev_block,
 		     unsigned long event, void *ptr);
@@ -39,6 +38,8 @@ void evl_net_put_dev(struct net_device *dev);
 
 void evl_net_wake_rx(struct net_device *dev);
 
+int __evl_net_dev_allocfd(struct net_device *dev);
+
 int evl_net_dev_allocfd(struct net *net, const char *devname);
 
 enum evl_net_rx_action
diff --git a/include/uapi/evl/net/device-abi.h b/include/uapi/evl/net/device-abi.h
index 7c2a2a206..fb2a30de7 100644
--- a/include/uapi/evl/net/device-abi.h
+++ b/include/uapi/evl/net/device-abi.h
@@ -12,5 +12,6 @@
 #define EVL_NETDEV_IOCBASE  0xef
 
 #define EVL_NDEVIOC_SETRXEBPF	_IOW(EVL_NETDEV_IOCBASE, 0, __s32 /* fd */)
+#define EVL_NDEVIOC_SWITCHOFF	_IO(EVL_NETDEV_IOCBASE, 1)
 
 #endif /* !_EVL_UAPI_NET_DEVICE_ABI_H */
diff --git a/include/uapi/evl/net/net-abi.h b/include/uapi/evl/net/net-abi.h
index 2dbdcbeba..bc1a5574d 100644
--- a/include/uapi/evl/net/net-abi.h
+++ b/include/uapi/evl/net/net-abi.h
@@ -13,11 +13,21 @@
 
 #define EVL_NET_IOCBASE  0xf0
 
+/* Turn on the oob port of a given device. */
+struct evl_net_devparams {
+	__u64 name_ptr;		/* (const char __user *name) */
+	__u64 poolsz;
+	__u64 bufsz;
+	__s32 fd;		/* [out] */
+};
+
+/* Get fildes on an open oob port. */
 struct evl_net_devfd {
 	__u64 name_ptr;		/* (const char __user *name) */
-	__u32 fd;
+	__s32 fd;		/* [out] */
 };
 
 #define EVL_NET_GETDEVFD	_IOWR(EVL_NETDEV_IOCBASE, 0, struct evl_net_devfd)
+#define EVL_NET_OPENPORT	_IOWR(EVL_NETDEV_IOCBASE, 1, struct evl_net_devparams)
 
 #endif /* !_EVL_UAPI_NET_NET_ABI_H */
diff --git a/include/uapi/evl/net/socket-abi.h b/include/uapi/evl/net/socket-abi.h
index 0393a81a1..e53aaacee 100644
--- a/include/uapi/evl/net/socket-abi.h
+++ b/include/uapi/evl/net/socket-abi.h
@@ -46,11 +46,6 @@ struct user_oob_msghdr {
 	struct __evl_timespec timeout;
 };
 
-struct evl_netdev_activation {
-	__u64 poolsz;
-	__u64 bufsz;
-};
-
 #define EVL_NEIGH_PERMANENT  0x1
 
 struct evl_net_solicit {
@@ -82,8 +77,6 @@ struct evl_net_sockopt {
 	__u64 optlen_ptr;		/* ([const] socklen_t __user *optlen) */
 };
 
-#define EVL_SOCKIOC_ACTIVATE	_IOW(EVL_SOCKET_IOCBASE, 1, struct evl_netdev_activation)
-#define EVL_SOCKIOC_DEACTIVATE	_IO(EVL_SOCKET_IOCBASE, 2)
 #define EVL_SOCKIOC_SENDMSG	_IOW(EVL_SOCKET_IOCBASE, 3, struct user_oob_msghdr)
 #define EVL_SOCKIOC_RECVMSG	_IOWR(EVL_SOCKET_IOCBASE, 4, struct user_oob_msghdr)
 #define EVL_SOCKIOC_SOLICIT	_IOW(EVL_SOCKET_IOCBASE, 5, struct evl_net_solicit)
diff --git a/kernel/evl/net/dev.c b/kernel/evl/net/dev.c
index 32c8fcd69..a93a534fc 100644
--- a/kernel/evl/net/dev.c
+++ b/kernel/evl/net/dev.c
@@ -27,6 +27,7 @@
 #include <evl/net/input.h>
 #include <evl/net/output.h>
 #include <evl/net/route.h>
+#include <uapi/evl/net/net-abi.h>
 
 /*
  * Since we need an EVL kthread to handle traffic from the out-of-band
@@ -45,13 +46,12 @@
 /*
  * The default number of I/O pages which should be available on a
  * per-device basis for conveying out-of-band traffic if not specified
- * by an EVL_SOCKIOC_ACTIVATE request.
+ * by an EVL_NET_DEVON request.
  */
 #define EVL_DEFAULT_NETDEV_POOLSZ  2048
 /*
  * The default fixed payload size available in I/O pages for conveying
- * out-of-band traffic through the device if not specified by an
- * EVL_SOCKIOC_ACTIVATE request.
+ * out-of-band traffic through the device.
  */
 #define EVL_DEFAULT_NETDEV_BUFSZ   PAGE_SIZE
 /*
@@ -102,7 +102,7 @@ start_handler_thread(struct net_device *dev,
  * or a VLAN interface.
  */
 static int enable_oob_port(struct net_device *dev,
-			struct evl_netdev_activation *act) /* inband, rtnl_lock held */
+			struct evl_net_devparams *p) /* inband, rtnl_lock held */
 {
 	struct oob_netdev_state *rnds, *nds;
 	struct evl_netdev_state *pest, *est;
@@ -141,19 +141,19 @@ static int enable_oob_port(struct net_device *dev,
 	if (est->refs++ > 0)	/* Guarded by rtnl_lock. */
 		goto queue;
 
-	if (!act->poolsz)
-		act->poolsz = EVL_DEFAULT_NETDEV_POOLSZ;
+	if (!p->poolsz)
+		p->poolsz = EVL_DEFAULT_NETDEV_POOLSZ;
 
-	if (!act->bufsz)
-		act->bufsz = EVL_DEFAULT_NETDEV_BUFSZ;
+	if (!p->bufsz)
+		p->bufsz = EVL_DEFAULT_NETDEV_BUFSZ;
 
 	/* Silently align on the current mtu if need be. */
 	mtu = READ_ONCE(real_dev->mtu);
-	if (act->bufsz < mtu)
-		act->bufsz = mtu;
+	if (p->bufsz < mtu)
+		p->bufsz = mtu;
 
-	est->pool_max = act->poolsz;
-	est->buf_size = act->bufsz;
+	est->pool_max = p->poolsz;
+	est->buf_size = p->bufsz;
 	spin_lock_init(&est->filter_lock);
 	est->qdisc = evl_net_alloc_qdisc(&evl_net_qdisc_fifo);
 	if (IS_ERR(est->qdisc)) {
@@ -320,7 +320,7 @@ static void disable_oob_port(struct net_device *dev) /* inband, rtnl_lock held *
 }
 
 static int switch_oob_port(struct net_device *dev,
-			struct evl_netdev_activation *act) /* rtnl_lock held, in-band */
+			struct evl_net_devparams *p) /* rtnl_lock held, in-band */
 {
 	int ret = 0;
 
@@ -329,39 +329,27 @@ static int switch_oob_port(struct net_device *dev,
 	 * received by the device flowing through the in-band net core
 	 * are diverted to netif_deliver_oob().
 	 */
-	if (act)
-		ret = enable_oob_port(dev, act);
-	else
+	if (p) {
+		if (p->poolsz > EVL_MAX_NETDEV_POOLSZ ||
+			p->bufsz > EVL_MAX_NETDEV_BUFSZ)
+			ret = -EINVAL;
+		else
+			ret = enable_oob_port(dev, p);
+	} else {
 		disable_oob_port(dev);
+	}
 
 	return ret;
 }
 
-/*
- * in-band, switches the oob port state of the device bound to the
- * socket.
- */
-int evl_net_switch_oob_port(struct evl_socket *esk,
-			struct evl_netdev_activation *act)
+int evl_net_switch_oob_port(struct net_device *dev,
+			struct evl_net_devparams *p)
 {
-	struct net_device *dev;
 	int ret;
 
-	dev = esk->proto->get_netif(esk);
-	if (dev == NULL)
-		return -ENXIO;
-
-	if (act &&
-		(act->poolsz > EVL_MAX_NETDEV_POOLSZ ||
-		act->bufsz > EVL_MAX_NETDEV_BUFSZ)) {
-		ret = -EINVAL;
-	} else {
-		rtnl_lock();
-		ret = switch_oob_port(dev, act);
-		rtnl_unlock();
-	}
-
-	evl_net_put_dev(dev);
+	rtnl_lock();
+	ret = switch_oob_port(dev, p);
+	rtnl_unlock();
 
 	return ret;
 }
@@ -470,12 +458,12 @@ void evl_net_put_dev(struct net_device *dev)
  */
 int netif_oob_switch_port(struct net_device *dev, bool enabled) /* rtnl_lock held */
 {
-	struct evl_netdev_activation act = {
+	struct evl_net_devparams p = {
 		.poolsz = 0,
 		.bufsz = 0,
 	};
 
-	return switch_oob_port(dev, enabled ? &act : NULL);
+	return switch_oob_port(dev, enabled ? &p : NULL);
 }
 
 /**
@@ -619,6 +607,9 @@ static long netdev_ioctl(struct file *filp, unsigned int cmd,
 	case EVL_NDEVIOC_SETRXEBPF:
 		ret = set_rx_filter(evl_net_real_dev(dev), arg);
 		break;
+	case EVL_NDEVIOC_SWITCHOFF:
+		ret = evl_net_switch_oob_port(dev, NULL);
+		break;
 	}
 
 	return ret;
@@ -642,16 +633,11 @@ static const struct file_operations netdev_fops = {
 #endif
 };
 
-int evl_net_dev_allocfd(struct net *net, const char *devname)
+int __evl_net_dev_allocfd(struct net_device *dev)
 {
-	struct net_device *dev;
 	struct file *filp;
 	int ret, fd;
 
-	dev = evl_net_get_dev_by_name(net, devname);
-	if (!dev)
-		return -EINVAL;
-
 	filp = anon_inode_getfile("[evl-netdev]", &netdev_fops,
 				dev, O_RDWR|O_CLOEXEC);
 	if (IS_ERR(filp))
@@ -673,3 +659,24 @@ int evl_net_dev_allocfd(struct net *net, const char *devname)
 
 	return ret;
 }
+
+int evl_net_dev_allocfd(struct net *net, const char *devname)
+{
+	struct net_device *dev;
+	int ret;
+
+	dev = evl_net_get_dev_by_name(net, devname);
+	if (!dev)
+		return -EINVAL;
+
+	ret = __evl_net_dev_allocfd(dev);
+	if (ret)
+		evl_net_put_dev(dev);
+
+	/*
+	 * On success, the reference on @dev will be released by
+	 * netdev_release().
+	 */
+
+	return ret;
+}
diff --git a/kernel/evl/net/net.c b/kernel/evl/net/net.c
index 972417dd0..9a0728f01 100644
--- a/kernel/evl/net/net.c
+++ b/kernel/evl/net/net.c
@@ -102,27 +102,55 @@ void __init evl_net_cleanup(void)
 static long net_ioctl(struct file *filp, unsigned int cmd,
 		unsigned long arg)
 {
-	struct evl_net_devfd req, __user *u_req;
+	struct evl_net_devparams devp, __user *u_devp;
+	struct evl_net_devfd fdreq, __user *u_fdreq;
+	struct net *net = current->nsproxy->net_ns;
 	const char __user *u_name;
 	struct filename *devname;
+	struct net_device *dev;
 	long ret;
 	int ufd;
 
 	switch (cmd) {
-	case EVL_NET_GETDEVFD:
-		u_req = (typeof(u_req))arg;
-		ret = copy_from_user(&req, u_req, sizeof(req));
+	case EVL_NET_OPENPORT: /* Turn oob port on. */
+		u_devp = (typeof(u_devp))arg;
+		ret = copy_from_user(&devp, u_devp, sizeof(devp));
+		if (ret)
+			return -EFAULT;
+		u_name = evl_valptr64(devp.name_ptr, const char);
+		devname = getname(u_name);
+		if (IS_ERR(devname))
+			return PTR_ERR(devname);
+		dev = dev_get_by_name(net, devname->name);
+		putname(devname);
+		if (!dev)
+			return -EINVAL;
+		ret = evl_net_switch_oob_port(dev, &devp);
+		dev_put(dev);	/* Drop the ref. obtained from dev_get_by_name() */
+		if (ret)
+			break;
+		/* The port is left open on user-specific errors. */
+		ufd = __evl_net_dev_allocfd(dev);
+		if (ufd < 0)
+			break;
+		ret = put_user((__u32)ufd, &u_devp->fd);
+		if (ret)
+			ret = -EFAULT;
+		break;
+	case EVL_NET_GETDEVFD:	/* Get a fildes on an oob-enabled device. */
+		u_fdreq = (typeof(u_fdreq))arg;
+		ret = copy_from_user(&fdreq, u_fdreq, sizeof(fdreq));
 		if (ret)
 			return -EFAULT;
-		u_name = evl_valptr64(req.name_ptr, const char);
+		u_name = evl_valptr64(fdreq.name_ptr, const char);
 		devname = getname(u_name);
 		if (IS_ERR(devname))
 			return PTR_ERR(devname);
-		ufd = evl_net_dev_allocfd(current->nsproxy->net_ns, devname->name);
+		ufd = evl_net_dev_allocfd(net, devname->name);
 		putname(devname);
 		if (ufd < 0)
 			return ufd;
-		ret = put_user((__u32)ufd, &u_req->fd);
+		ret = put_user((__u32)ufd, &u_fdreq->fd);
 		if (ret)
 			return -EFAULT;
 		break;
diff --git a/kernel/evl/net/socket.c b/kernel/evl/net/socket.c
index 38065222f..123ea4f30 100644
--- a/kernel/evl/net/socket.c
+++ b/kernel/evl/net/socket.c
@@ -813,9 +813,8 @@ __poll_t sock_oob_poll(struct file *filp,
 static long sock_inband_ioctl(struct sock *sk, unsigned int cmd,
 			unsigned long arg)
 {
-	struct evl_socket *esk = evl_sk(sk);
-	struct evl_netdev_activation act, __user *u_act;
 	struct evl_net_solicit solreq, __user *u_solreq;
+	struct evl_socket *esk = evl_sk(sk);
 	int ret;
 
 	/*
@@ -824,16 +823,6 @@ static long sock_inband_ioctl(struct sock *sk, unsigned int cmd,
 	 * request from the inband stage for those would fail.
 	 */
 	switch (cmd) {
-	case EVL_SOCKIOC_ACTIVATE: /* Turn oob port on. */
-		u_act = (typeof(u_act))arg;
-		ret = raw_copy_from_user(&act, u_act, sizeof(act));
-		if (ret)
-			return -EFAULT;
-		ret = evl_net_switch_oob_port(esk, &act);
-		break;
-	case EVL_SOCKIOC_DEACTIVATE: /* Turn oob port off. */
-		ret = evl_net_switch_oob_port(esk, NULL);
- 		break;
 	case EVL_SOCKIOC_SOLICIT:
 		u_solreq = (typeof(u_solreq))arg;
 		ret = copy_from_user(&solreq, u_solreq, sizeof(solreq));
-- 
2.30.2

